### ISSUE TITLE
Breaking Bug! Get selected rows returns a null row?

### ISSUE THREAD

**Issue by AmitMY** [ceolter/ag-grid#973]
This is breaking many options in my code, and I guess in many others (Every bulk action on a table data)
It seems the bug was introduced from 4.2.X, though I cannot confirm.

Selecting a row, when the grid only has one row, says that the grid has 2 rows selected:
[Object, null]

Here is a plunker: http://plnkr.co/edit/b0NM91hBVm0o1uVS8HW8?p=preview
(Check console for selected rows)

Please note, that it is not just for a grid with one row. It happens even when a grid has X rows and is filtered to one row.

**Comment by ceolter** [comment]
thanks for raising. your right, there is an extra node getting selected.

in the last release, an extra 'parent' node was introduced, which was work towards the new pivoting functionality (refactoring we did, in anticipation of pivoting).

i'm looking into it, see if it can be easily fixed.

**Comment by AmitMY** [comment]
Thanks
I don't fully understand why it happens, even with your explanation..
Hope you can fix it in a patch update, for now I went through all the places I use getSelectedRows and filtered the null values out..

**Comment by ceolter** [comment]
btw, i don't check github on a daily basis, if i have email notification, i jsut get to many emails

if something is a bad bug, mention it on the members forum

**Comment by AmitMY** [comment]
K, will do next time!

**Comment by ceolter** [comment]
ok, i've already fixed this.

not sure if you want to know, but what i was trying to explain above is there is now a 'parent of all rows' - so even if you are not grouping, there is always one group, and the parent of that group is never rendered.

so if you have a grid with 100 rows, there is a 'rootNode' that has these 100 rows as children, and this rootNode is never displayed. if you are looking at grid code, you will see this rootNode in InMemoryRowController, it's silently owning all the other nodes.

anyway, to fix, i need to update selectionController method onRowSelected - it's already fixed in latest in master - but you can't use this, i'll have to put into a branch and release off branch, otherwise you get all new stuff we are working on that's not properly tested.

the code is now this:
```
    private onRowSelected(event: any): void {
        var rowNode = event.node;

        if (rowNode.isSelected()) {
            this.selectedNodes[rowNode.id] = rowNode;
        } else {
            this.selectedNodes[rowNode.id] = undefined;
        }
    }
```

and it needs to be

```
    private onRowSelected(event: any): void {
        var rowNode = event.node;

        // we do not store the group rows when the groups select children
        if (this.groupSelectsChildren && rowNode.group) { return; }

        if (rowNode.isSelected()) {
            this.selectedNodes[rowNode.id] = rowNode;
        } else {
            this.selectedNodes[rowNode.id] = undefined;
        }
    }
```

basically when you have groupSelectsChildren=true, then the groups also potentially get selected. the magic entry you see in the selected nodes is that rootNode that is getting selected when you select a row.

i don't want to do a release now as it's 9:30pm here - i need to be awake for a few hours after a do a patch in case i mess it up. so i'll do a patch in the morning. if you want to hack the grid code with the above one line entry you can tell me if it solves everything (ie help me test!!!)

**Comment by ceolter** [comment]
fixed and released. can you verify?

**Comment by AmitMY** [comment]
Yea it is fixed!
Thank you so much

I now understand why it happened, and that makes me wonder - if everything is a group, this must be possible - https://github.com/ceolter/ag-grid/issues/974
(Becasue every group calculates aggregated data, and even without me creating a group, it seems like it calls the aggFunc by itself)

Also, can you please permit my user to the enterprise forum as well?
https://ag-grid.com/forum/member.php?action=profile&uid=1708

**Comment by ceolter** [comment]
access to members granted. please go easy on us, keep the issues short and to the point :)

**Comment by AmitMY** [comment]
Yeah of course, I usually do
